src/main.rs

```diff
@@ -1,35 +1,209 @@
 
 fn show_info() {
     eprintln!("MUSI-6106 Assignment Executable");
     eprintln!("(c) 2024 Stephen Garrett & Ian Clester");
 }
 
 fn main() {
-   show_info();
+    show_info();
 
     // Parse command line arguments
     let args: Vec<String> = std::env::args().collect();
-    if args.len() < 3 {
-        eprintln!("Usage: {} <input wave filename> <output text filename>", args[0]);
+    if args.len() < 6 {
+        eprintln!("Usage: {} <input wav filename> <output wav filename> <FIR/IIR> <gain> <delay in seconds>", args[0]);
         return
     }
 
-    // Open the input wave file
-    let mut reader = hound::WavReader::open(&args[1]).unwrap();
+    let input_path = &args[1];
+    let output_path = &args[2];
+    let filter_type = match args[3].as_str() {
+        "FIR" => FilterType::FIR,
+        "IIR" => FilterType::IIR,
+        _ => {
+            eprintln!("Invalid filter type. Choose 'FIR' or 'IIR'.");
+            std::process::exit(1);
+        },
+    };
+    let gain: f32 = args[4].parse().unwrap();
+    let delay_secs: f32 = args[5].parse().unwrap();
+
+
+    let input_file = File::open(input_path).unwrap();
+    let mut reader = WavReader::new(input_file).unwrap();
     let spec = reader.spec();
-    let channels = spec.channels;
 
-    // TODO: Modify this to process audio in blocks using your comb filter and write the result to an audio file.
-    //       Use the following block size:
-    let block_size = 1024;
+    let output_file = File::create(output_path).unwrap();
+    let mut writer = WavWriter::new(BufWriter::new(output_file), spec).unwrap();
+
+    let sample_rate_hz = spec.sample_rate as f32;
+    let channels = spec.channels as usize;
+
+    let mut comb_filter = CombFilter::new(filter_type, delay_secs, sample_rate_hz, channels);
+    comb_filter.set_param(FilterParam::Gain, gain).unwrap();
+
+    let block_size = 1024; 
+    let samples_per_block = block_size * channels;
+
+    while let Ok(block) = reader.samples::<i16>().take(samples_per_block).collect::<Result<Vec<_>, _>>() {
+        if block.is_empty() {
+            break;
+        }
+        let mut processed_samples = vec![0f32; block.len()];
+        let block_samples_f32: Vec<f32> = block.iter().map(|&s| s as f32 / i16::MAX as f32).collect();
+    
+        for channel in 0..channels {
+            let mut channel_samples = Vec::with_capacity(block.len() / channels);
+            for (i, sample) in block_samples_f32.iter().enumerate() {
+                if i % channels == channel {
+                    channel_samples.push(*sample);
+                }
+            }
+    
+            let mut processed_channel_samples = vec![0f32; channel_samples.len()];
+            comb_filter.process(&[&channel_samples], &mut [&mut processed_channel_samples]);
+    
+            for (i, &sample) in processed_channel_samples.iter().enumerate() {
+                processed_samples[i * channels + channel] = sample;
```

Comment:
Could rename this &sample variable to &samples so that it's easier to see that there's more than one value.
```suggestion
            for (i, &samples) in processed_channel_samples.iter().enumerate() {
                processed_samples[i * channels + channel] = samples;
```